src/garage/tf/policies/categorical_mlp_policy.py

```diff
@@ -1,103 +1,146 @@
+
+        self.model = MLPModel(output_dim=self.action_dim,
+                              hidden_sizes=hidden_sizes,
+                              hidden_nonlinearity=hidden_nonlinearity,
+                              hidden_w_init=hidden_w_init,
+                              hidden_b_init=hidden_b_init,
+                              output_nonlinearity=output_nonlinearity,
+                              output_w_init=output_w_init,
+                              output_b_init=output_b_init,
+                              layer_normalization=layer_normalization,
+                              name='MLPModel')
+
+        self._initialize()
+
+    def _initialize(self):
+        state_input = tf.compat.v1.placeholder(tf.float32,
+                                               shape=(None, self.obs_dim))
+
+        with tf.compat.v1.variable_scope(self.name) as vs:
+            self._variable_scope = vs
+            self.model.build(state_input)
+
+        self._f_prob = tf.compat.v1.get_default_session().make_callable(
+            self.model.networks['default'].outputs,
+            feed_list=[self.model.networks['default'].input])
 
     @property
     def vectorized(self):
+        """Vectorized or not."""
         return True
 
     @overrides
     def dist_info_sym(self, obs_var, state_info_vars=None, name=None):
-        with tf.name_scope(name, 'dist_info_sym', [obs_var, state_info_vars]):
-            with tf.name_scope(self._prob_network_name, values=[obs_var]):
-                prob = L.get_output(
-                    self._l_prob, {self._l_obs: tf.cast(obs_var, tf.float32)})
-            return dict(prob=prob)
+        """Symbolic graph of the distribution."""
+        with tf.compat.v1.variable_scope(self._variable_scope):
+            prob = self.model.build(obs_var, name=name)
+        return dict(prob=prob)
 
     @overrides
     def dist_info(self, obs, state_infos=None):
-        return dict(prob=self._f_prob(obs))
+        """Distribution info."""
+        prob = self._f_prob(obs)
+        return dict(prob=prob)
 
-    # The return value is a pair. The first item is a matrix (N, A), where each
-    # entry corresponds to the action value taken. The second item is a vector
-    # of length N, where each entry is the density value for that action, under
-    # the current policy
     @overrides
     def get_action(self, observation):
+        """Return a single action."""
         flat_obs = self.observation_space.flatten(observation)
         prob = self._f_prob([flat_obs])[0]
         action = self.action_space.weighted_sample(prob)
         return action, dict(prob=prob)
 
     def get_actions(self, observations):
+        """Return multiple actions."""
         flat_obs = self.observation_space.flatten_n(observations)
         probs = self._f_prob(flat_obs)
         actions = list(map(self.action_space.weighted_sample, probs))
         return actions, dict(prob=probs)
 
+    def get_regularizable_vars(self):
```

Comment:
Should we add this function to the base class as well?

Reply:
`CNN` and `GRU` policies will have different implementation of this function since they use different naming convention for parameters. I think we should leave this here.